src/shims/tls.rs

```diff
@@ -142,12 +172,15 @@ impl<'tcx> TlsData<'tcx> {
         for (&key, TlsEntry { data, dtor }) in
             thread_local.range_mut((start, Unbounded))
         {
-            if let Some(data_scalar) = *data {
-                if let Some(dtor) = dtor {
-                    let ret = Some((*dtor, data_scalar, key));
-                    *data = None;
-                    return ret;
+            match data.entry(thread_id) {
+                Entry::Occupied(entry) => {
+                    let data_scalar = entry.remove();
```

Comment:
There is a subtle change in behavior here... previously we only set `data` to `None` if there was a `dtor`, now we do it unconditionally.

Going from the comment above this function (that I copied from somewhere), I think we should only set the data to NULL if there actually is a dtor.